src/dagmc/DagMC.cpp

```diff
@@ -0,0 +1,774 @@
+  ErrorCode rval;
+
+  // write out a mesh file if requested
+  if (ffile && 0 < flen) {
+    rval = MBI->write_mesh(ffile);
+    if (MB_SUCCESS != rval) {
+      std::cerr << "Failed to write mesh to " << ffile << "." << std::endl;
+      return rval;
+    }
+  }
+
+  return MB_SUCCESS;
+}
+
+/* SECTION V: Metadata handling */
+
+ErrorCode DagMC::get_group_name(EntityHandle group_set, std::string& name) {
+  ErrorCode rval;
+  const void* v = NULL;
+  int ignored;
+  rval = MBI->tag_get_by_ptr(name_tag(), &group_set, 1, &v, &ignored);
+  if (MB_SUCCESS != rval)
+    return rval;
+  name = static_cast<const char*>(v);
+  return MB_SUCCESS;
+}
+
+ErrorCode DagMC::parse_group_name(EntityHandle group_set, prop_map& result, const char* delimiters) {
+  ErrorCode rval;
+  std::string group_name;
+  rval = get_group_name(group_set, group_name);
+  if (rval != MB_SUCCESS)
+    return rval;
+
+  std::vector< std::string > group_tokens;
+  tokenize(group_name, group_tokens, delimiters);
+
+  // iterate over all the keyword positions
+  // keywords are even indices, their values (optional) are odd indices
+  for (unsigned int i = 0; i < group_tokens.size(); i += 2) {
+    std::string groupkey = group_tokens[i];
+    std::string groupval;
+    if (i < group_tokens.size() - 1)
+      groupval = group_tokens[i + 1];
+    result[groupkey] = groupval;
+  }
+  return MB_SUCCESS;
+}
+
+ErrorCode DagMC::detect_available_props(std::vector<std::string>& keywords_list, const char* delimiters) {
+  ErrorCode rval;
+  std::set< std::string > keywords;
+  for (std::vector<EntityHandle>::const_iterator grp = group_handles().begin();
+       grp != group_handles().end(); ++grp) {
+    std::map< std::string, std::string > properties;
+    rval = parse_group_name(*grp, properties, delimiters);
+    if (rval == MB_TAG_NOT_FOUND)
+      continue;
+    else if (rval != MB_SUCCESS)
+      return rval;
+
+    for (prop_map::iterator i = properties.begin();
+         i != properties.end(); ++i) {
+      keywords.insert((*i).first);
+    }
+  }
+  keywords_list.assign(keywords.begin(), keywords.end());
+  return MB_SUCCESS;
+}
+
+ErrorCode DagMC::append_packed_string(Tag tag, EntityHandle eh,
+                                      std::string& new_string) {
+  // When properties have multiple values, the values are tagged in a single character array
+  // with the different values separated by null characters
+  ErrorCode rval;
+  const void* p;
+  const char* str;
+  int len;
+  rval = MBI->tag_get_by_ptr(tag, &eh, 1, &p, &len);
+  if (rval == MB_TAG_NOT_FOUND) {
+    // This is the first entry, and can be set directly
+    p = new_string.c_str();
+    return MBI->tag_clear_data(tag, &eh, 1, p, new_string.length() + 1);
+  } else if (rval != MB_SUCCESS)
+    return rval;
+  else {
+    str = static_cast<const char*>(p);
+  }
+
+  // append a new value for the property to the existing property string
+  unsigned int tail_len = new_string.length() + 1;
```

Comment:
Readability: If the `+ 1` is being appended here to compensate for the `NULL` at the end of `char* new_packed_string`, it would be slightly more readable to calculate `new_len` here, explicitly add `+ 1`, then declare `new_packed_string` as `new char[ new_len ]`